From 908d8a4fd19794ff30a10ea994094f0b919362f1 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Wed, 19 Mar 2025 09:50:49 +0100
Subject: [PATCH] media: stm32: csi: ensure to configure all lanes in
 lane-merger

All lanes routing should be configured within the lane-merger
register even if they are not enabled. Update the lane-merger
configuration code to ensure that all lanes routing is done and
all have a unique value.

Change-Id: I2e10635dd89c640c195ac9debdedb3fa6c3ce97c
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/445146
Tested-by: Alain VOLMAT <alain.volmat@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
---
 drivers/media/platform/st/stm32/stm32-csi.c | 26 +++++++++++++++------
 1 file changed, 19 insertions(+), 7 deletions(-)

--- a/drivers/media/platform/st/stm32/stm32-csi.c
+++ b/drivers/media/platform/st/stm32/stm32-csi.c
@@ -366,19 +366,31 @@ struct stm32_csi_dev *v4l2_subdev_to_csi
 
 static int stm32_csi_setup_lane_merger(struct stm32_csi_dev *csi2priv)
 {
-	int i;
+	int i, j;
 	u32 lmcfgr;
+	u32 lanes_used = 0;
 
 	lmcfgr = readl_relaxed(csi2priv->base + STM32_CSI_LMCFGR);
 	lmcfgr &= ~(STM32_CSI_LMCFGR_LANENB_MASK | STM32_CSI_LMCFGR_DL0MAP_MASK |
 		    STM32_CSI_LMCFGR_DL1MAP_MASK);
 
-	for (i = 0; i < csi2priv->num_lanes; i++) {
-		/* Check that lane ID is < max number of lane */
-		if (!csi2priv->lanes[i] || csi2priv->lanes[i] > STM32_CSI_LANES_MAX) {
-			dev_err(csi2priv->dev, "Invalid lane id (%d)\n",
-				csi2priv->lanes[i]);
-			return -EINVAL;
+	/* We need to route all lanes, even if not enabled later on */
+	for (i = 0; i < STM32_CSI_LANES_MAX; i++) {
+		if (i < csi2priv->num_lanes) {
+			/* Check that lane ID is not 0 and < max number of lane */
+			if (!csi2priv->lanes[i] || csi2priv->lanes[i] > STM32_CSI_LANES_MAX) {
+				dev_err(csi2priv->dev, "Invalid lane id (%d)\n",
+					csi2priv->lanes[i]);
+				return -EINVAL;
+			}
+			lanes_used |= BIT(csi2priv->lanes[i]);
+		} else {
+			for (j = 1; j <= STM32_CSI_LANES_MAX; j++) {
+				if (!(lanes_used & BIT(j))) {
+					csi2priv->lanes[i] = j;
+					lanes_used |= BIT(j);
+				}
+			}
 		}
 		lmcfgr |= (csi2priv->lanes[i] << ((i * 4) +
 			   STM32_CSI_LMCFGR_DL0MAP_SHIFT));
